pandas/core/arraylike.py

```diff
@@ -157,11 +157,67 @@ def array_ufunc(self, ufunc: Callable, method: str, *inputs: Any, **kwargs: Any)
     --------
     numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_ufunc__
     """
+    from pandas.core.frame import DataFrame
     from pandas.core.generic import NDFrame
     from pandas.core.internals import BlockManager
 
     cls = type(self)
 
+    is_ndframe = [isinstance(x, NDFrame) for x in inputs]
```

Comment:
move this entire thing to a separate function and just call it here. This function is already completely out of control.

Reply:
Good idea, will move to a helper function

Reply:
yes i see that great